sstables/sstable_set.cc

```diff
@@ -0,0 +1,526 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <vector>
+#include <seastar/core/shared_ptr.hh>
+#include "sstables.hh"
+#include "schema.hh"
+#include "sstable_set.hh"
+#include "compatible_ring_position.hh"
+#include <boost/range/adaptors.hpp>
+
+namespace sstables {
+
+sstable_run::sstable_run(std::unordered_set<shared_sstable>&& sstables)
+    : _all(sstables) {
+}
+void sstable_run::insert(shared_sstable sst) {
+    _all.insert(std::move(sst));
+}
+
+void sstable_run::erase(shared_sstable sst) {
+    _all.erase(sst);
+}
+
+uint64_t sstable_run::data_size() const {
+    return boost::accumulate(_all | boost::adaptors::transformed(std::mem_fn(&sstable::data_size)), uint64_t(0));
+}
+
+std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
+    os << "Run = {\n";
+    if (run.all().empty()) {
+        os << "  Identifier: not found\n";
+    } else {
+        os << format("  Identifier: {}\n", (*run.all().begin())->run_identifier());
+    }
+
+    auto frags = boost::copy_range<std::vector<shared_sstable>>(run.all());
+    boost::sort(frags, [] (const shared_sstable& x, const shared_sstable& y) {
+        return x->get_first_decorated_key().token() < y->get_first_decorated_key().token();
+    });
+    os << "  Fragments = {\n";
+    for (auto& frag : frags) {
+        os << format("    {}={}:{}\n", frag->generation(), frag->get_first_decorated_key().token(), frag->get_last_decorated_key().token());
+    }
+    os << "  }\n}\n";
+    return os;
+}
+
+
+sstable_list::sstable_list() = default;
+sstable_list::sstable_list(lw_shared_ptr<const std::unordered_set<versioned_sstable>> versioned_sstables, uint64_t size, uint64_t version)
+    : _versioned_sstables(std::move(versioned_sstables))
+    , _unversioned_sstables(make_lw_shared<std::unordered_set<shared_sstable>>())
```

Comment:
Why are you changing sstable_list?

Reply:
The sstable_list is the return type of sstable_set::all(). It used to be an lw_shared_ptr<unordered_set<shared_sstable>> but it can't be any more because we don't have a separate unordered_set of all sstables for every version.

Reply:
Makes sense. I did not review the new sstable_list, so the following may be completely wrong, but shouldn't it be just a pair of iterators into a version? That is, now that taking a snapshot is very cheap, an sstable_list can hold a reference to a version, and begin()/end() can call the sstable_set's implementations.

Reply:
Ideally, that should be it, but in one case the sstable_list was not a reference to sstable_sets all(), but a modified copy of it, so in order not to copy everything in that case, I've added a modifications list to it.
When we solve the non-linear version issue, the local modifications list will not be needed any more, because instead of keeping a modified copy of all sstables from a snapshot, we will be able to simply keep a reference to a modified snapshot.

Reply:
If we want a fast size() method, we will still need to cache it though, but it could be stored in the version, not the sstable_list